src/main/scala/mesosphere/marathon/core/storage/repository/impl/GroupRepositoryImpl.scala

```diff
@@ -0,0 +1,184 @@
+      .build()
+  }
+}
+
+object StoredGroup {
+  val DateFormat = DateTimeFormatter.ISO_OFFSET_DATE_TIME
+
+  def apply(group: Group): StoredGroup =
+    StoredGroup(
+      id = group.id,
+      appIds = group.apps.mapValues(_.version.toOffsetDateTime),
+      storedGroups = group.groups.map(StoredGroup(_))(collection.breakOut),
+      dependencies = group.dependencies,
+      version = group.version.toOffsetDateTime)
+
+  def apply(proto: Protos.StoredGroup): StoredGroup = {
+    val apps: Map[PathId, OffsetDateTime] = proto.getAppIdsList.asScala.map { appId =>
+      PathId.fromSafePath(appId.getId()) -> OffsetDateTime.parse(appId.getVersion, DateFormat)
+    }(collection.breakOut)
+
+    val groups = proto.getGroupsList.asScala.map(StoredGroup(_))
+
+    StoredGroup(
+      id = PathId.fromSafePath(proto.getId),
+      appIds = apps,
+      storedGroups = groups.toVector,
+      dependencies = proto.getDependenciesList.asScala.map(PathId.fromSafePath)(collection.breakOut),
+      version = OffsetDateTime.parse(proto.getVersion, DateFormat)
+    )
+  }
+}
+
+class StoredGroupRepositoryImpl[K, C, S](
+    persistenceStore: PersistenceStore[K, C, S],
+    protected val appRepository: AppRepository)(
+    implicit
+    ir: IdResolver[PathId, StoredGroup, C, K],
+    marshaller: Marshaller[StoredGroup, S],
+    unmarshaller: Unmarshaller[S, StoredGroup],
+    val ctx: ExecutionContext,
+    val mat: Materializer
+) extends GroupRepository {
+  import StoredGroupRepositoryImpl._
+
+  /* TODO: We're bypassing the cache so it won't store StoredGroup, but we should likely cache
+    the root. We should probably have read-after-write consistency...
+    */
+
+  private val storedRepo = {
+    def leafStore(store: PersistenceStore[K, C, S]): PersistenceStore[K, C, S] = store match {
+      case s: BasePersistenceStore[K, C, S] => s
+      case s: LoadTimeCachingPersistenceStore[K, C, S] => leafStore(s.store)
+      case s: LazyCachingPersistenceStore[K, C, S] => leafStore(s.store)
+    }
+    new PersistenceStoreVersionedRepository[PathId, StoredGroup, K, C, S](leafStore(persistenceStore), _.id, _.version)
+  }
+
+  override def root(): Future[Group] = async {
+    val unresolved = await(storedRepo.get(RootId))
+    unresolved.map(_.resolve(this, appRepository)) match {
+      case Some(group) =>
+        await(group)
+      case None =>
+        Group.empty
+    }
+  }
+
+  override def rootVersions(): Source[OffsetDateTime, NotUsed] =
+    storedRepo.versions(RootId)
+
+  override def rootVersion(version: OffsetDateTime): Future[Option[Group]] = async {
+    val unresolved = await(storedRepo.getVersion(RootId, version))
+    unresolved.map(_.resolve(this, appRepository)) match {
+      case Some(group) =>
+        Some(await(group))
+      case None =>
+        None
+    }
+  }
+
+  /**
+    * @inheritdoc
+    * @todo We need to eventually garbage collect as we will accumulate apps/appVersions that neither
```

Comment:
We already have a couple of actors that run from time to time to do all kinds of gc operations. We can easily add another one.